pkg/util/retry/retry.go

```diff
@@ -159,6 +165,9 @@ func WithMaxAttempts(ctx context.Context, opts Options, n int, fn func() error)
 			return nil
 		}
 	}
+	if err == nil {
+		err = errors.Errorf("function never gets run")
```

Comment:
Can we make this error message a little clearer? For example, `errors.Wrap("did not run function", ctx.Err())`